File: src/rust/src/lite/ffi.rs
//
// Copyright 2019-2022 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

//! For passing values to/from another language.

#[cfg(any(target_os = "ios", feature = "check-all"))]
pub mod ios {
    use libc::size_t;

    pub trait FromOrDefault<S>: From<S> + Default {
        fn from_or_default(maybe: Option<S>) -> Self;
    }

    impl<S, T: From<S> + Default> FromOrDefault<S> for T {
        fn from_or_default(maybe: Option<S>) -> Self {
            maybe.map(Self::from).unwrap_or_default()
        }
    }

    #[repr(C)]
    #[derive(Debug, Default)]
    pub struct rtc_OptionalU16 {
        pub value: u16,
        pub valid: bool,
    }

    impl From<u16> for rtc_OptionalU16 {
        fn from(value: u16) -> Self {
            Self { value, valid: true }
        }
    }

    /// Swift "UInt32?"
    #[repr(C)]
    #[derive(Debug, Default)]
    pub struct rtc_OptionalU32 {
        pub value: u32,
        pub valid: bool,
    }

    impl From<u32> for rtc_OptionalU32 {
        fn from(value: u32) -> Self {
            Self { value, valid: true }
        }
    }

    /// Swift Data/[UInt8]/UnsafeBufferPointer<UInt8>
    #[repr(C)]
    #[derive(Debug)]
    pub struct rtc_Bytes<'a> {
        pub ptr: *const u8,
        pub count: size_t,
        phantom: std::marker::PhantomData<&'a u8>,
    }

    impl<'a> Default for rtc_Bytes<'a> {
        fn default() -> Self {
            Self {
                ptr: std::ptr::null(),
                count: 0,
                phantom: std::marker::PhantomData,
            }
        }
    }

    impl<'a, T: AsRef<[u8]> + ?Sized> From<&'a T> for rtc_Bytes<'a> {
        fn from(bytes: &'a T) -> Self {
            let bytes = bytes.as_ref();
            Self {
                ptr: bytes.as_ptr(),
                count: bytes.len(),
                phantom: std::marker::PhantomData,
            }
        }
    }

    impl<'a> rtc_Bytes<'a> {
        pub fn as_slice(&self) -> &[u8] {
            if self.ptr.is_null() {
                return &[];
            }
            unsafe { std::slice::from_raw_parts(self.ptr, self.count) }
        }

        pub fn to_vec(&self) -> Vec<u8> {
            self.as_slice().to_vec()
        }
    }

    /// Swift String
    #[repr(C)]
    #[derive(Debug)]
    pub struct rtc_String<'a> {
        pub ptr: *const u8,
        pub count: size_t,
        phantom: std::marker::PhantomData<&'a u8>,
    }

    impl<'a> Default for rtc_String<'a> {
        fn default() -> Self {
            Self {
                ptr: std::ptr::null(),
                count: 0,
                phantom: std::marker::PhantomData,
            }
        }
    }

    impl<'a, T: AsRef<str> + ?Sized> From<&'a T> for rtc_String<'a> {
        fn from(s: &'a T) -> Self {
            let s = s.as_ref();
            Self {
                ptr: s.as_ptr(),
                count: s.len(),
                phantom: std::marker::PhantomData,
            }
        }
    }

    impl<'a> rtc_String<'a> {
        pub fn as_str(&self) -> Option<&str> {
            if self.ptr.is_null() {
                return None;
            }
            let bytes = unsafe { std::slice::from_raw_parts(self.ptr, self.count) };
            std::str::from_utf8(bytes).ok()
        }

        pub fn to_string(&self) -> Option<String> {
            Some(self.as_str()?.to_string())
        }
    }
}

File: src/ios/SignalRingRTC/SignalRingRTC/FFI.swift
//
// Copyright 2019-2022 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalRingRTC.RingRTC

// Here is the new lite/modularized pattern for FFI for a module called X
// (so far where X is "sfu" or "http")
// 1. A swift XClient in X.swift owns a rust X::Client in x.rs
//    which it interacts with via a pointer to X::Client
//    and C funcs prefixed with "rtc_x_Client".
// 2. A rust X::Client owns an RC to a swift XDelegateWrapper
//    which wraps the XDelegate.  The X::Clients get the pointer
//    to the XDelegateWrapper and C funcs from a struct
//    called rtc_x_Delegate that is passed into rtc_x_Client_create.
//    It uses those to make calls to the XDelegateWrapper
//    which then call the XDelegate.
// 3. The swift XDelegate is implemented by the client application.
// 4. All C module-specific structs are defined in x.rs under the x::ios module
//    and have the prefix "rtx_x_".  Common C structs are defined here.
//
// Sometimes a module doesn't need all these parts, but that's the
// general idea.

extension rtc_OptionalU16 {
    func asUInt16() -> UInt16? {
        if self.valid {
            return self.value
        } else{
            return nil
        }
    }
}

extension rtc_OptionalU32 {
    func asUInt32() -> UInt32? {
        if self.valid {
            return self.value
        } else{
            return nil
        }
    }
}

extension rtc_Bytes {
    static func empty() -> Self {
        return Self(ptr: nil, count: 0)
    }

    // Don't forget to call deallocate()
    static func allocate(from maybe: [UInt8]?) -> Self {
        guard let array = maybe else {
            return Self.empty()
        }

        let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: array.count)
        ptr.initialize(from: array, count: array.count)
        return Self(ptr: ptr, count: array.count)
    }

    // Don't forget to call deallocate()
    static func allocate(from maybe: Data?) -> Self {
        guard let data = maybe else {
            return Self.empty()
        }

        return Self.allocate(from: Array(data))
    }

    func deallocate() {
        if self.ptr == nil {
            return
        }
        self.ptr.deallocate()
    }

    func toArray() -> [UInt8]? {
        if self.ptr == nil {
            return nil
        }
        return Array(self.asUnsafeBufferPointer())
    }

    func toData() -> Data? {
        if self.ptr == nil {
            return nil
        }
        return Data(self.asUnsafeBufferPointer())
    }

    func asUnsafeBufferPointer() -> UnsafeBufferPointer<UInt8> {
        return UnsafeBufferPointer(start: self.ptr, count: self.count)
    }

    func toUUID() -> UUID? {
        guard let ptr = UnsafeRawPointer(self.ptr),
              self.count >= MemoryLayout<uuid_t>.size else {
            return nil
        }
        return UUID(uuid: ptr.loadUnaligned(as: uuid_t.self))
    }
}

extension ContiguousBytes {
    func withRtcBytes<R>(_ body: (rtc_Bytes) throws -> R) rethrows -> R {
        return try withUnsafeBytes { buffer in
            let bytes = rtc_Bytes(ptr: buffer.baseAddress?.assumingMemoryBound(to: UInt8.self), count: buffer.count)
            return try body(bytes)
        }
    }
}

extension rtc_String {
    static func empty() -> Self {
        return Self(ptr: nil, count: 0)
    }

    // Don't forget to call deallocate()
    static func allocate(from maybe: String?) -> Self {
        guard let string = maybe else {
            return Self.empty()
        }

        let bytes = Array(string.utf8);
        let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: bytes.count)
        ptr.initialize(from: bytes, count: bytes.count)
        return Self(ptr: ptr, count: bytes.count)
    }

    func deallocate() {
        if self.ptr == nil {
            return
        }
        self.ptr.deallocate()
    }

    func toString() -> String? {
        if self.ptr == nil {
            return nil
        }
        return String(bytes: UnsafeBufferPointer(start: self.ptr, count: self.count), encoding: .utf8)
    }
}

class Requests<T> {
    private var continuationById: [UInt32: CheckedContinuation<T, Never>] = [:]
    private var nextId: UInt32 = 1

    func add(_ continuation: CheckedContinuation<T, Never>) -> UInt32 {
        let id = self.nextId
        self.nextId &+= 1
        self.continuationById[id] = continuation
        return id
    }

    func resolve(id: UInt32, response: T) -> Bool {
        if let continuation = self.continuationById.removeValue(forKey: id) {
            continuation.resume(returning: response)
            return true
        }
        return false
    }
}
